MAINT: Avoid BytesWarning in PyArray_DescrConverter()

jdufresne opened the pull request:
Fixes #9308

jdufresne committed:
MAINT: Avoid all BytesWarning

A BytesWarning can be emitted when bytes are and strings are mismatched.
Catching BytesWarning ensures a better boundary between str and bytes
type. The test suite is now run with the -b flag to emit this warning.

Fixes #9308

mattip commented:
LGTM. Could you add a test to prevent such regressions in the future? Maybe our travis test run that does `RUN_FULL_TESTS=1` could also add `-b` in `tools/travis-test.sh ?

charris commented:
Needs a test.

seberg commented:
@jdufresne can you add a test? (I think there was one in the issue). Also it would be good to add an immediate failure path where Eric commented. I am not sure, but on first sight, that can likely be triggered easily by using `"śä".encode("utf8")` or similar as a key.

jdufresne commented on numpy/core/src/multiarray/descriptor.c:
```diff
@@ -1664,16 +1663,21 @@ PyArray_DescrConverter(PyObject *obj, PyArray_Descr **at)
+            if (PyBytes_Check(obj)) {
                 PyObject *tmp;
                 tmp = PyUnicode_FromEncodedObject(obj, "ascii", "strict");
                 if (tmp != NULL) {
```
Changing this to propagate the error will change the following:

```python
np.dtype, b'\xff'
```

To raise a `UnicodeDecodeError: 'ascii' codec can't decode byte 0xff in position 0: ordinal not in range(128)` instead of the previous `TypeError`. Is that ok? This is codified in the test `test_dtype_from_bytes`. I'm happy to fix this if that is indeed the preference.

jdufresne replied:
I opened #14591 so this can be reviewed and analyzed for correctness in isolation to this change.

seberg replied:
@jdufresne that example segfaults on my NumPy 1.16.2, so I do not think we have to worry about it too much. The TypeError I get on master is actually nicer then a decode error. I am not sure how fine it is to set a new exception while one is already on the fly... Replacing the `goto error;` with `goto fail;` will not change the behaviour but make it clear that the error is being handled... Sorry that I did not realize that before.

jdufresne committed:
MAINT: Explicitly goto fail when PyUnicode_FromEncodedObject() return NULL

Makes the code a bit more explicit and readable in intent.

jdufresne commented:
Ok, I've changed the error handling to `goto fail` to revert back to the `TypeError`.

I also added `-b` to the Travis and tox test suite now fail whenever there is a mismatch between `str` and `bytes`. Adding this in caught a few other mishandling as well so I've fixed them as well. Let me know what you think. Thanks.

seberg commented:
Lets put this in, thanks @jdufresne, also for the other cleanups!
